File: android/core/src/main/java/ru/maksimbulva/chess/core/engine/move/generator/KnightMoveGenerator.kt
package ru.maksimbulva.chess.core.engine.move.generator

import ru.maksimbulva.chess.core.engine.board.Cell
import ru.maksimbulva.chess.core.engine.board.Vector2
import ru.maksimbulva.chess.core.engine.move.Move
import ru.maksimbulva.chess.core.engine.position.Position

internal object KnightMoveGenerator : BaseDeltasMoveGenerator() {

    private val KNIGHT_DELTAS = arrayOf(
        Vector2(-1, -2),
        Vector2(1, -2),
        Vector2(-2, -1),
        Vector2(2, -1),
        Vector2(-2, 1),
        Vector2(2, 1),
        Vector2(-1, 2),
        Vector2(1, 2)
    )

    override fun generateMoves(fromCell: Cell, position: Position, moves: MutableList<Move>) {
        return generateMoves(fromCell, position, KNIGHT_DELTAS, moves)
    }
}

File: ChessApp/ChessEngine/Move/Generator/KnightMoveGenerator.cs
namespace ChessEngine.Move.Generator
{
    internal sealed class KnightMoveGenerator : BaseDeltasMoveGenerator
    {
        private static readonly MoveDelta[] KnightDeltas = new MoveDelta[]
        {
            new MoveDelta(-1, -2),
            new MoveDelta(1, -2),
            new MoveDelta(-2, -1),
            new MoveDelta(2, -1),
            new MoveDelta(-2, 1),
            new MoveDelta(2, 1),
            new MoveDelta(-1, 2),
            new MoveDelta(1, 2)
        };

        protected override MoveDelta[] GetMoveDeltas() => KnightDeltas;

        protected override Piece GetPiece() => Piece.Knight;
    }
}
